Readme.md

# Calendar

  Calendar UI component designed for use as a date-picker,
  full-sized calendar or anything in-between.

  ![javascript calendar component](http://f.cl.ly/items/2u3w1D421W0C370Z3G1U/Screen%20Shot%202012-10-11%20at%2014.32.41.png)

  Live demo is [here](http://component.github.io/calendar/)

## Installation

    $ component install component/calendar

## Example

```js
var Calendar = require('calendar');
var cal = new Calendar;
cal.el.appendTo('body');
```

## Events

  - `view change` (date, action) when the viewed month/year is changed without modification of the selected date. This can be done either by next/prev buttons or dropdown menu. The action will be "prev", "next", "month" or "year" depending on what action caused the view to change.
  - `change` (date) when the selected date is modified

## API

### new Calendar(date)

  Initialize a new `Calendar` with the given `date` shown,
  defaulting to now.

### Calendar#select(date)

  Select the given `date` (`Date` object).

### Calendar#show(date)

  Show the given `date`. This does _not_ select the given date,
  it simply ensures that it is visible in the current view.

### Calendar#showMonthSelect()

  Add month selection input.

### Calendar#showYearSelect([from], [to])

  Add year selection input, with optional range specified by `from` and `to`,
  which default to the current year -10 / +10.

### Calendar#prev()

  Show the previous view (month).

### Calendar#next()

  Show the next view (month).

### Calendar#min()

  Define earliest valid date - calendar won't generate `change` events for dates before this one.

### Calendar#max()

  Define latest valid date - calendar won't generate `change` events for dates after this one.

### Calendar#locale({months, weekdaysMin})

  Set alternative locale:
  - `months` - an array of 12 strings representing month names _January..December_.
  - `weekdaysMin` - an array of 7 strings representing day names shortcuts _Sunday..Saturday_

## Themes

  [Aurora](https://github.com/component/aurora-calendar):

  ![](http://f.cl.ly/items/043N1r0e1L130y162R2f/Screen%20Shot%202012-09-17%20at%209.17.32%20PM.png)

# License

  MIT



History.md

0.2.0 / 2015-04-27
==================

 * enable changing locales for calendar
 * use component templates to translate HTML

0.1.0 / 2014-04-02
==================

 * add dates restriction: min/max functions

0.0.5 / 2013-06-06 
==================

 * fix tags

0.0.4 / 2013-05-31
==================

 * pin deps
 * fix short months selection on 31th of the month

0.0.3 / 2012-09-21
==================

  * fix start of the month date padding [colinf]

0.0.2 / 2012-09-20
==================

  * add package.json
  * fix prev/next month day selection [colinf]


test/dayrange.js
var DayRange = require('../lib/dayrange')
  , assert = require('assert');

describe('day range', function(){
  it('should consider all dates as valid if no min/max specified', function(){
    var dr = new DayRange;
    assert.ok(!dr.before(new Date));
    assert.ok(!dr.after(new Date));
    assert.ok(dr.valid([2002, 12, 10]));
  });

  it('should consider dates inside of the range as valid', function(){
    var dr = new DayRange([2014, 3, 2], [2014, 4, 3]);
    assert.ok(dr.before([2014, 3, 1]));
    assert.ok(!dr.valid([2014, 3, 1]));
    assert.ok(dr.valid([2014, 3, 2]));
    assert.ok(dr.valid([2014, 3, 30]));
    assert.ok(dr.valid([2014, 4, 3]));
    assert.ok(!dr.valid([2014, 4, 4]));
    assert.ok(dr.after([2014, 4, 4]));
  });

  it('should work with mixture of dates and arrays', function(){
    var dr = new DayRange()
      .min([2014, 3, 2])
      .max(new Date(2014, 4, 3));
    assert.ok(dr.before(new Date(2014, 3, 1)));
    assert.ok(!dr.valid(new Date(2014, 3, 1)));
    assert.ok(dr.valid(new Date(2014, 3, 2)));
    assert.ok(dr.valid(new Date(2014, 3, 30)));
    assert.ok(dr.valid(new Date(2014, 4, 3)));
    assert.ok(!dr.valid(new Date(2014, 4, 4)));
    assert.ok(dr.after(new Date(2014, 4, 4)));
  });

  it('should work if only min is specified', function(){
    var dr = new DayRange([2013, 3, 3]);
    assert.ok(!dr.valid([2013, 3, 2]));
    assert.ok(dr.valid([2013, 3, 3]));
    assert.ok(dr.valid([2013, 3, 4]));
  });

  it('should work if only max is specified', function(){
    var dr = new DayRange(null, [2013, 3, 3]);
    assert.ok(dr.valid([2013, 3, 2]));
    assert.ok(dr.valid([2013, 3, 3]));
    assert.ok(!dr.valid([2013, 3, 4]));
  });
});

lib/calendar.js

/**
 * Module dependencies.
 */

var bind = require('bind')
  , domify = require('domify')
  , Emitter = require('emitter')
  , classes = require('classes')
  , Days = require('./days');

/**
 * Expose `Calendar`.
 */

module.exports = Calendar;

/**
 * Initialize a new `Calendar`
 * with the given `date` defaulting
 * to now.
 *
 * Events:
 *
 *  - `prev` when the prev link is clicked
 *  - `next` when the next link is clicked
 *  - `change` (date) when the selected date is modified
 *
 * @params {Date} date
 * @api public
 */

function Calendar(date) {
  if (!(this instanceof Calendar)) {
    return new Calendar(date);
  }

  Emitter.call(this);
  var self = this;
  this.el = domify('<div class=calendar></div>');
  this.days = new Days;
  this.el.appendChild(this.days.el);
  this.on('change', bind(this, this.show));
  this.days.on('prev', bind(this, this.prev));
  this.days.on('next', bind(this, this.next));
  this.days.on('year', bind(this, this.menuChange, 'year'));
  this.days.on('month', bind(this, this.menuChange, 'month'));
  this.show(date || new Date);
  this.days.on('change', function(date){
    self.emit('change', date);
  });
}

/**
 * Mixin emitter.
 */

Emitter(Calendar.prototype);

/**
 * Add class `name` to differentiate this
 * specific calendar for styling purposes,
 * for example `calendar.addClass('date-picker')`.
 *
 * @param {String} name
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.addClass = function(name){
  classes(this.el).add(name);
  return this;
};

/**
 * Select `date`.
 *
 * @param {Date} date
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.select = function(date){
  if (this.days.validRange.valid(date)) {
    this.selected = date;
    this.days.select(date);
  }
  this.show(date);
  return this;
};

/**
 * Show `date`.
 *
 * @param {Date} date
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.show = function(date){
  this._date = date;
  this.days.show(date);
  return this;
};

/**
 * Set minimum valid date (inclusive)
 *
 * @param {Date} date
 * @api public
 */

Calendar.prototype.min = function(date) {
  this.days.validRange.min(date);
  return this;
};


/**
 * Set maximum valid date (inclusive)
 *
 * @param {Date} date
 * @api public
 */

Calendar.prototype.max = function(date) {
  this.days.validRange.max(date);
  return this;
};

/**
 * Enable a year dropdown.
 *
 * @param {Number} from
 * @param {Number} to
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.showYearSelect = function(from, to){
  from = from || this._date.getFullYear() - 10;
  to = to || this._date.getFullYear() + 10;
  this.days.yearMenu(from, to);
  this.show(this._date);
  return this;
};

/**
 * Enable a month dropdown.
 *
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.showMonthSelect = function(){
  this.days.monthMenu();
  this.show(this._date);
  return this;
};

/**
 * Return the previous month.
 *
 * @return {Date}
 * @api private
 */

Calendar.prototype.prevMonth = function(){
  var date = new Date(this._date);
  date.setDate(1);
  date.setMonth(date.getMonth() - 1);
  return date;
};

/**
 * Return the next month.
 *
 * @return {Date}
 * @api private
 */

Calendar.prototype.nextMonth = function(){
  var date = new Date(this._date);
  date.setDate(1);
  date.setMonth(date.getMonth() + 1);
  return date;
};

/**
 * Show the prev view.
 *
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.prev = function(){
  this.show(this.prevMonth());
  this.emit('view change', this.days.selectedMonth(), 'prev');
  return this;
};

/**
 * Show the next view.
 *
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.next = function(){
  this.show(this.nextMonth());
  this.emit('view change', this.days.selectedMonth(), 'next');
  return this;
};

/**
 * Switch to the year or month selected by dropdown menu.
 *
 * @return {Calendar}
 * @api public
 */

Calendar.prototype.menuChange = function(action){
  var date = this.days.selectedMonth();
  this.show(date);
  this.emit('view change', date, action);
  return this;
};

/**
 * Select locale
 *
 * @param {months, weekdaysMin} - array of months labels and array of weekdays shortcuts
 * @api public
 */

 Calendar.prototype.locale = function(locale) {
  this.days.setLocale(locale);
  this.days.show(this._date);
  return this;
 };


lib/days.js

/**
 * Module dependencies.
 */

var domify = require('domify')
  , Emitter = require('emitter')
  , classes = require('classes')
  , events = require('event')
  , map = require('map')
  , template = require('./template.html')
  , inGroupsOf = require('in-groups-of')
  , clamp = require('./utils').clamp
  , range = require('range')
  , empty = require('empty')
  , normalize = require('normalize')
  , DayRange = require('./dayrange');

/**
 * Default locale - en
 */

var LOCALE = {
  months: 'January_February_March_April_May_June_July_August_September_October_November_December'.split('_'),
  weekdaysMin: 'Su_Mo_Tu_We_Th_Fr_Sa'.split('_')
};


/**
 * Get days in `month` for `year`.
 *
 * @param {Number} month
 * @param {Number} year
 * @return {Number}
 * @api private
 */

function daysInMonth(month, year) {
  return [31, (isLeapYear(year) ? 29 : 28), 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month];
}

/**
 * Check if `year` is a leap year.
 *
 * @param {Number} year
 * @return {Boolean}
 * @api private
 */

function isLeapYear(year) {
  return (0 === year % 400)
    || ((0 === year % 4) && (0 !== year % 100))
    || (0 === year);
}


/**
 * Expose `Days`.
 */

module.exports = Days;

/**
 * Initialize a new `Days` view.
 *
 * Emits:
 *
 *   - `prev` when prev link is clicked
 *   - `next` when next link is clicked
 *   - `change` (date) when a date is selected
 *
 * @api public
 */

function Days() {
  Emitter.call(this);
  var self = this;
  this.el = domify(template);
  classes(this.el).add('calendar-days');
  this.head = this.el.tHead;
  this.body = this.el.tBodies[0];
  this.title = this.head.querySelector('.title');
  this.select(new Date);
  this.validRange = new DayRange;
  this.locale = LOCALE;

  // emit "day"
  events.bind(this.body, 'click', normalize(function(e){
    if (e.target.tagName !== 'A') {
      return true;
    }

    e.preventDefault();

    var el = e.target;
    var data = el.getAttribute('data-date').split('-');
    if (!self.validRange.valid(data)) {
      return false;
    }
    var year = data[0];
    var month = data[1];
    var day = data[2];
    var date = new Date(year, month, day);
    self.select(date);
    self.emit('change', date);
    return false;
  }));

  // emit "prev"
  events.bind(this.el.querySelector('.prev'), 'click', normalize(function(ev){
    ev.preventDefault();

    self.emit('prev');
    return false;
  }));

  // emit "next"
  events.bind(this.el.querySelector('.next'), 'click', normalize(function(ev){
    ev.preventDefault();

    self.emit('next');
    return false;
  }));
}

/**
 * Mixin emitter.
 */

Emitter(Days.prototype);

/**
 * Select the given `date`.
 *
 * @param {Date} date
 * @return {Days}
 * @api public
 */

Days.prototype.select = function(date){
  this.selected = date;
  return this;
};


/**
 * Select locale
 *
 * @param {months, weekdaysMin} - array of months labels and array of weekdays shortcuts
 * @api public
 */

 Days.prototype.setLocale = function(locale) {
  this.locale = locale;
  return this;
 };

/**
 * Show date selection.
 *
 * @param {Date} date
 * @api public
 */

Days.prototype.show = function(date){
  var year = date.getFullYear();
  var month = date.getMonth();
  this.showSelectedYear(year);
  this.showSelectedMonth(month);
  var subhead = this.head.querySelector('.subheading');
  if (subhead) {
    subhead.parentElement.removeChild(subhead);
  }

  this.head.appendChild(this.renderHeading(this.locale.weekdaysMin));
  empty(this.body);
  this.body.appendChild(this.renderDays(date));
};

/**
 * Enable a year dropdown.
 *
 * @param {Number} from
 * @param {Number} to
 * @api public
 */

Days.prototype.yearMenu = function(from, to){
  this.selectYear = true;
  this.title.querySelector('.year').innerHTML = yearDropdown(from, to);
  var self = this;
  events.bind(this.title.querySelector('.year .calendar-select'), 'change', function(){
    self.emit('year');
    return false;
  });
};

/**
 * Enable a month dropdown.
 *
 * @api public
 */

Days.prototype.monthMenu = function(){
  this.selectMonth = true;
  this.title.querySelector('.month').innerHTML = monthDropdown(this.locale.months);
  var self = this;
  events.bind(this.title.querySelector('.month .calendar-select'), 'change', function(){
    self.emit('month');
    return false;
  });
};

/**
 * Return current year of view from title.
 *
 * @api private
 */

Days.prototype.titleYear = function(){
  if (this.selectYear) {
    return this.title.querySelector('.year .calendar-select').value;
  } else {
    return this.title.querySelector('.year').innerHTML;
  }
};

/**
 * Return current month of view from title.
 *
 * @api private
 */

Days.prototype.titleMonth = function(){
  if (this.selectMonth) {
    return this.title.querySelector('.month .calendar-select').value;
  } else {
    return this.title.querySelector('.month').innerHTML;
  }
};

/**
 * Return a date based on the field-selected month.
 *
 * @api public
 */

Days.prototype.selectedMonth = function(){
  return new Date(this.titleYear(), this.titleMonth(), 1);
};

/**
 * Render days of the week heading with
 * the given `length`, for example 2 for "Tu",
 * 3 for "Tue" etc.
 *
 * @param {String} len
 * @return {Element}
 * @api private
 */

Days.prototype.renderHeading = function(days){
  var rows = '<tr class=subheading>' + map(days, function(day){
    return '<th>' + day + '</th>';
  }).join('') + '</tr>';
  return domify(rows);
};

/**
 * Render days for `date`.
 *
 * @param {Date} date
 * @return {Element}
 * @api private
 */

Days.prototype.renderDays = function(date){
  var rows = this.rowsFor(date);
  var html = map(rows, function(row){
    return '<tr>' + row.join('') + '</tr>';
  }).join('\n');
  return domify(html);
};

/**
 * Return rows array for `date`.
 *
 * This method calculates the "overflow"
 * from the previous month and into
 * the next in order to display an
 * even 5 rows.
 *
 * @param {Date} date
 * @return {Array}
 * @api private
 */

Days.prototype.rowsFor = function(date){
  var selected = this.selected;
  var selectedDay = selected.getDate();
  var selectedMonth = selected.getMonth();
  var selectedYear = selected.getFullYear();
  var month = date.getMonth();
  var year = date.getFullYear();

  // calculate overflow
  var start = new Date(date);
  start.setDate(1);
  var before = start.getDay();
  var total = daysInMonth(month, year);
  var perRow = 7;
  var totalShown = perRow * Math.ceil((total + before) / perRow);
  var after = totalShown - (total + before);
  var cells = [];

  // cells before
  cells = cells.concat(cellsBefore(before, month, year, this.validRange));

  // current cells
  for (var i = 0; i < total; ++i) {
    var day = i + 1
      , select = (day == selectedDay && month == selectedMonth && year == selectedYear);
    cells.push(renderDay([year, month, day], this.validRange, select));
  }

  // after cells
  cells = cells.concat(cellsAfter(after, month, year, this.validRange));

  return inGroupsOf(cells, 7);
};

/**
 * Update view title or select input for `year`.
 *
 * @param {Number} year
 * @api private
 */

Days.prototype.showSelectedYear = function(year){
  if (this.selectYear) {
    this.title.querySelector('.year .calendar-select').value = year;
  } else {
    this.title.querySelector('.year').innerHTML = year;
  }
};

/**
 * Update view title or select input for `month`.
 *
 * @param {Number} month
 * @api private
 */

Days.prototype.showSelectedMonth = function(month) {
  if (this.selectMonth) {
    this.title.querySelector('.month .calendar-select').value = month;
  } else {
    this.title.querySelector('.month').innerHTML = this.locale.months[month];
  }
};

/**
 * Return `n` days before `month`.
 *
 * @param {Number} n
 * @param {Number} month
 * @return {Array}
 * @api private
 */

function cellsBefore(n, month, year, validRange){
  var cells = [];
  if (month === 0) --year;
  var prev = clamp(month - 1);
  var before = daysInMonth(prev, year);
  while (n--) cells.push(renderDay([year, prev, before--], validRange, false, 'prev-day'));
  return cells.reverse();
}

/**
 * Return `n` days after `month`.
 *
 * @param {Number} n
 * @param {Number} month
 * @return {Array}
 * @api private
 */

function cellsAfter(n, month, year, validRange){
  var cells = [];
  var day = 0;
  if (month == 11) ++year;
  var next = clamp(month + 1);
  while (n--) cells.push(renderDay([year, next, ++day], validRange, false, 'next-day'));
  return cells;
}


/**
 * Day template.
 */

function renderDay(ymd, validRange, selected, style) {
  var date = 'data-date=' + ymd.join('-')
    , styles = []
    , tdClass = ''
    , aClass = '';

  if (selected) {
    tdClass = ' class="selected"';
  }
  if (style) {
    styles.push(style);
  }
  if (!validRange.valid(ymd)) {
    styles.push('invalid');
  }
  if (styles.length) {
    aClass = ' class="' + styles.join(' ') + '"';
  }


  return '<td' + tdClass + '><a ' + date + aClass + '>' + ymd[2] + '</a></td>';
}

/**
 * Year dropdown template.
 */

function yearDropdown(from, to) {
  var years = range(from, to, 'inclusive');
  var options = map(years, yearOption).join('');
  return '<select class="calendar-select">' + options + '</select>';
}

/**
 * Month dropdown template.
 */

function monthDropdown(months) {
  var options = map(months, monthOption).join('');
  return '<select class="calendar-select">' + options + '</select>';
}

/**
 * Year dropdown option template.
 */

function yearOption(year) {
  return '<option value="' + year + '">' + year + '</option>';
}

/**
 * Month dropdown option template.
 */

function monthOption(month, i) {
  return '<option value="' + i + '">' + month + '</option>';
}
